src/datasets/formatting/formatting.py

```diff
@@ -154,26 +159,44 @@ def extract_batch(self, pa_table: pa.Table) -> dict:
         return {col: self._arrow_array_to_numpy(pa_table[col]) for col in pa_table.column_names}
 
     def _arrow_array_to_numpy(self, pa_array: pa.Array) -> np.ndarray:
-        zero_copy_only = _is_zero_copy_only(pa_array.type)
         if isinstance(pa_array, pa.ChunkedArray):
-            # don't call to_numpy() directly or we end up with a np.array with dtype object
-            # call to_numpy on the chunks instead
-            # for ArrayExtensionArray call py_list directly to support dynamic dimensions
             if isinstance(pa_array.type, _ArrayXDExtensionType):
-                array: List = [row for chunk in pa_array.chunks for row in chunk.to_pylist()]
+                # don't call to_pylist() to preserve dtype of the fixed-size array
+                zero_copy_only = _is_zero_copy_only(pa_array.type.storage_dtype, unnest=True)
+                if pa_array.type.shape[0] is None:
+                    array: List = [
+                        row
+                        for chunk in pa_array.chunks
+                        for row in chunk.to_list_of_numpy(zero_copy_only=zero_copy_only)
+                    ]
+                else:
+                    array: List = [
+                        row for chunk in pa_array.chunks for row in chunk.to_numpy(zero_copy_only=zero_copy_only)
+                    ]
             else:
+                zero_copy_only = _is_zero_copy_only(pa_array.type) and all(
+                    not _is_array_with_nulls(chunk) for chunk in pa_array.chunks
+                )
                 array: List = [
                     row for chunk in pa_array.chunks for row in chunk.to_numpy(zero_copy_only=zero_copy_only)
                 ]
         else:
-            # cast to list of arrays or we end up with a np.array with dtype object
-            # for ArrayExtensionArray call py_list directly to support dynamic dimensions
             if isinstance(pa_array.type, _ArrayXDExtensionType):
-                array: List = pa_array.to_pylist()
+                # don't call to_pylist() to preserve dtype of the fixed-size array
+                zero_copy_only = _is_zero_copy_only(pa_array.type.storage_dtype, unnest=True)
+                if pa_array.type.shape[0] is None:
+                    array: List = pa_array.to_list_of_numpy(zero_copy_only=zero_copy_only)
+                else:
+                    array: List = pa_array.to_numpy(zero_copy_only=zero_copy_only)
```

Comment:
If I understand correctly the idea is to always return numpy arrays, even for the default formatting of a dataset.

I think this is a (very welcomed) breaking change, feel free to mention it in this PR's description if it's not already

Reply:
This is explained in the comment: `# don't call to_pylist() to preserve dtype of the fixed-size array`.
This change is needed because previously we would call `pa_array.to_pylist()`, which in turn calls `pa_array.to_numpy(zero_copy_only=zero_copy_only).tolist()` for arrays where the first dimension is not dynamic, and the `tolist()` call is dangerous because after that call we can't restore the initial dtype properly. So what I did here is I copied `the ArrayExtensionArray.to_pylist` code and removed the `to_list` call for arrays without dynamic dims. (So this should not be a breaking change.)

Reply:
Oh yes you're right ! For some reason I thought it would affect the default python formatting x)